src/config/index.ts

```diff
@@ -0,0 +1,130 @@
+import jsPlugin from '@eslint/js';
+import type { ESLint, Linter } from 'eslint';
+import prettierConfig from 'eslint-config-prettier';
+import importPlugin from 'eslint-plugin-import';
+import reactPlugin from 'eslint-plugin-react';
+import reactHooksPlugin from 'eslint-plugin-react-hooks';
+import globals from 'globals';
+import tsEslint from 'typescript-eslint';
+import rules from '../rules';
+
+export type ConfigOptions = Readonly<
+  Partial<{
+    /**
+     * TypeScript の設定ファイル (languageOptions.parserOptions.project)
+     * デフォルト: ./tsconfig.json
+     */
+    tsProject: boolean | string | string[] | undefined;
+    /**
+     * React を使用するか. true の場合, React に関連する設定を有効にする
+     * デフォルト: false
+     */
+    react: boolean | undefined;
+    /**
+     * Prettier を使用するか. true の場合, eslint-config-prettier を使ってフォーマットに関するルールを無効化する
+     * デフォルト: true
+     */
+    prettier: boolean | undefined;
+  }>
+>;
+
+/**
+ * ESLint の設定を作る
+ * @param options オプション
+ * @param configs カスタム設定の配列
+ * @returns 設定の配列
+ */
+export function config(options?: ConfigOptions, configs?: readonly Linter.FlatConfig[]): Linter.FlatConfig[] {
+  const tsProject = options?.tsProject ?? './tsconfig.json';
+  const react = options?.react ?? false;
+  const prettier = options?.prettier ?? true;
+
+  return [
+    // # Linter 自体の設定
+    {
+      plugins: {
+        '@typescript-eslint': tsEslint.plugin as ESLint.Plugin,
+        'import': importPlugin,
+        'react': reactPlugin,
+        'react-hooks': reactHooksPlugin,
+      },
+    },
+    // # 言語ごとの構文や実行環境などの設定
+    {
+      files: ['**/*.{js,jsx,cjs,mjs}', '**/*.{ts,tsx,cts,mts}'],
+      languageOptions: {
+        // ES2020 の構文がパースできるように。
+        // NOTE: アプリケーションによってサポートされている ECMAScript のバージョンは違うので、
+        // 本来であればアプリケーションで利用している Node.js のバージョンやサポートブラウザの
+        // バージョンに合わせて上書きするべきするべき設定だが、いちいち上書きするのも面倒なので、
+        // ひとまず一番最新のバージョンが利用可能であるとしておき、必要に応じてアプリケーションサイドで
+        // 上書きしてもらう、という運用にする
+        ecmaVersion: 2020,
+        parserOptions: react ? { ecmaFeatures: { jsx: true } } : {},
+        globals: {
+          // `ecmaVersion` に揃えておく
+          ...globals.es2020,
+          ...(react ? globals.browser : {}),
+        },
+      },
+      settings: {
+        react: { version: 'detect' },
+      },
+    },
+    {
+      files: ['**/*.{js,jsx,mjs}'],
+      languageOptions: {
+        // 現代では type="script" な環境で JS を書くことはまずないので、
+        // デフォルトで type="module" なJSであるとして lint する
+        sourceType: 'module',
+      },
+    },
+    {
+      files: ['**/*.cjs'],
+      languageOptions: {
+        sourceType: 'commonjs',
+      },
+    },
+    {
+      files: ['**/*.{ts,tsx,cts,mts}'],
+      languageOptions: {
+        sourceType: 'module',
+        parser: tsEslint.parser as Linter.FlatConfigParserModule,
+        parserOptions: tsProject ? { project: tsProject } : {},
```

Comment:
MAY: It seems that `tsProject  === false` or `tsProject === ''` will result in `{}`. Is that intended?

Wouldn't the following code be better?

```suggestion
        parserOptions: tsProject ?? { project: tsProject },,
```

Reply:
It's intended. You can unset `parserOptions.project` by setting a falsy value if you don't use type-aware rules.
(The config enables type-aware rules by default, so it's very uncommon though.)

Maybe we can disable type-aware rules when `tsProject` is false in the future version (as I do in [my own config](https://github.com/susisu/eslint-config/blob/08cbfbc0dff8151b56fa24aaaeddbd8e27cfd340/src/config/index.ts#L96)).